# Total Production Units for Self-Consumption

### Master in Data Science and Engineering - FEUP

#### Group 4
202107955 - Beatriz Iara Nunes Silva
\
202206252 - Inês Clotilde da Costa Neves
\
202502527 - Kirill Savin
\
202502528 - Mariana Rocha Cristino
\
202202895 - Patrícia Crespo da Silva

<div id="top"></div>

# Table of Content

<ol>
    <li><h3><a href="#introduction">Introduction</a></h3></li>
    <li><h3><a href="#research">Research Questions</a></h3></li>
    <li><h3><a href="#data">Data</a></h3></li>
    <li><h3><a href="#inference">Inferences</a></h3></li>
    <li><h3><a href="#results">Results and Discussion</a></h3></li>
    <li><h3><a href="#conclusion">Conclusion</a></h3></li>    

</ol>

# Introduction

<div id="research"></div>
<strong><a href="#top">Back to top</a></strong>

# Research Questions

**General Research Question:**
* RQ: Compare how seasonal (winter vs summer), regional, and technical factors shape self-consumption energy production patterns in Portugal between 2023 and 2024.

**Specifics Research Questions:**
* RQ1: Compare the average installed capacity per UPAC across different power levels and districts in 2023 and 2024. - Kirill
* RQ2: Compare the evolution of installed capacity between 2023 and 2024 across residential and industrial UPACs to assess differences in growth patterns. - Iara and Mariana
* RQ3: Compare the total installed capacity for self-consumption across different power scales (installed capacity ranges) and seasons (winter vs. summer) in selected Portuguese districts during 2023 and 2024. - Inês and Patrícia

Districts for SRQ: Aveiro, Évora, Vila Real and Faro

<div id="research"></div>
<strong><a href="#top">Back to top</a></strong>

# Data

## Libraries

In [1]:
import geopandas as gpd
import json
from matplotlib.cm import ScalarMappable
from matplotlib.colors import Normalize
import matplotlib.colors as mcolors
import matplotlib.gridspec as gridspec
import matplotlib.pyplot as plt
from matplotlib import patheffects
import numpy as np
import pandas as pd
import plotly.express as px
import plotly.graph_objects as go
from plotly.subplots import make_subplots
import scipy.stats as stats
import seaborn as sns
import statsmodels.api as sm
import statsmodels

## Prepare Data

Reading the dataset:

In [2]:
df_origin = pd.read_csv('../Data/UPAC_Total_Production.csv', sep=';', decimal='.')
df_origin.head(10)

Quarter District     Municipality                              Parish  \
0  2023T1  Coimbra  Condeixa-a-Nova                           Furadouro   
1  2023T1  Coimbra  Condeixa-a-Nova                            Zambujal   
2  2023T1  Coimbra  Condeixa-a-Nova  Condeixa-a-Velha e Condeixa-a-Nova   
3  2023T1  Coimbra  Condeixa-a-Nova               Vila Seca e Bem da Fé   
4  2023T1  Coimbra  Figueira da Foz                           São Pedro   
5  2023T1  Coimbra  Figueira da Foz                  Moinhos da Gândara   
6  2023T1  Coimbra  Figueira da Foz                Buarcos e São Julião   
7  2023T1  Coimbra  Figueira da Foz                Buarcos e São Julião   
8  2023T1  Coimbra  Figueira da Foz                     Ferreira-a-Nova   
9  2023T1  Coimbra  Figueira da Foz                               Lavos   

   Zip Code Technology Type Voltage level Installed power range (kW)  \
0      3150           Solar           BTN                     ]0, 4]   
1      3150           Solar           BTN                     ]0, 4]   
2      3150   Não Atribuído           BTN                     ]0, 4]   
3      3150           Solar           BTN                     ]0, 4]   
4      3090   Não Atribuído           BTN                     ]0, 4]   
5      3080   Não Atribuído           BTN                     ]0, 4]   
6      3080   Não Atribuído           BTN                     ]0, 4]   
7      3080           Solar           BTN                     ]0, 4]   
8      3090   Não Atribuído           BTN                     ]0, 4]   
9      3090           Solar           BTN                     ]0, 4]   

   Number of installations  Total installed power (kW)  DistrictCode  \
0                        2                        3.00             6   
1                        2                        4.32             6   
2                        1                        1.05             6   
3                       17                       28.14             6   
4                        2                        3.28             6   
5                        1                        2.00             6   
6                        9                       13.59             6   
7                      128                      202.51             6   
8                        3                        4.96             6   
9                       93                      112.51             6   

   Municipality Code DistrictMunicipalityParishCode  CPEs (#)  \
0                604                         060407    9537.0   
1                604                         060410    9537.0   
2                604                         060411    9537.0   
3                604                         060413    9537.0   
4                605                         060514   50436.0   
5                605                         060518   50436.0   
6                605                         060520   50436.0   
7                605                         060520   50436.0   
8                605                         060521   50436.0   
9                605                         060522   50436.0   

   relacao_instalacoes_por_cpe  relacao_potencia_por_cpe  
0                     0.000210                  0.000315  
1                     0.000210                  0.000453  
2                     0.000105                  0.000110  
3                     0.001783                  0.002951  
4                     0.000040                  0.000065  
5                     0.000020                  0.000040  
6                     0.000178                  0.000269  
7                     0.002538                  0.004015  
8                     0.000059                  0.000098  
9                     0.001844                  0.002231

Rename the last two columns to English

In [3]:
df_origin = df_origin.rename(columns={
    "relacao_instalacoes_por_cpe": "installations_per_cpe_ratio",
    "relacao_potencia_por_cpe": "power_per_cpe_ratio"
})
print(df_origin.columns)

Index(['Quarter', 'District', 'Municipality', 'Parish', 'Zip Code',
       'Technology Type', 'Voltage level', 'Installed power range (kW)',
       'Number of installations', 'Total installed power (kW)', 'DistrictCode',
       'Municipality Code', 'DistrictMunicipalityParishCode', 'CPEs (#)',
       'installations_per_cpe_ratio', 'power_per_cpe_ratio'],
      dtype='object')


Dataset info

In [4]:
print("\nDataset info:")
print(df_origin.info())


Dataset info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 121294 entries, 0 to 121293
Data columns (total 16 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   Quarter                         121294 non-null  object 
 1   District                        121294 non-null  object 
 2   Municipality                    121294 non-null  object 
 3   Parish                          121294 non-null  object 
 4   Zip Code                        121294 non-null  int64  
 5   Technology Type                 121283 non-null  object 
 6   Voltage level                   121292 non-null  object 
 7   Installed power range (kW)      121294 non-null  object 
 8   Number of installations         121294 non-null  int64  
 9   Total installed power (kW)      121294 non-null  float64
 10  DistrictCode                    121294 non-null  int64  
 11  Municipality Code               121294 non-null  int64  
 12  D

Non-Numeric Columns

In [5]:
non_numeric = df_origin.select_dtypes(exclude=['number'])


for col in non_numeric.columns:
    unique_vals = df_origin[col].unique().tolist()
    print(f"Column: {col} — {len(unique_vals)} unique values")
    print(unique_vals)
    print("-" * 40)

Column: Quarter — 11 unique values
['2023T1', '2023T2', '2024T2', '2024T3', '2024T4', '2023T3', '2023T4', '2024T1', '2025T1', '2025T2', '2022T4']
----------------------------------------
Column: District — 18 unique values
['Coimbra', 'Évora', 'Faro', 'Guarda', 'Leiria', 'Lisboa', 'Portalegre', 'Porto', 'Santarém', 'Setúbal', 'Viana do Castelo', 'Vila Real', 'Viseu', 'Aveiro', 'Beja', 'Braga', 'Bragança', 'Castelo Branco']
----------------------------------------
Column: Municipality — 278 unique values
['Condeixa-a-Nova', 'Figueira da Foz', 'Lousã', 'Mira', 'Miranda do Corvo', 'Montemor-o-Velho', 'Oliveira do Hospital', 'Pampilhosa da Serra', 'Penacova', 'Penela', 'Soure', 'Tábua', 'Vila Nova de Poiares', 'Alandroal', 'Arraiolos', 'Borba', 'Estremoz', 'Évora', 'Montemor-o-Novo', 'Mora', 'Portel', 'Redondo', 'Reguengos de Monsaraz', 'Vendas Novas', 'Viana do Alentejo', 'Vila Viçosa', 'Albufeira', 'Alcoutim', 'Aljezur', 'Castro Marim', 'Faro', 'Lagoa', 'Lagos', 'Loulé', 'Monchique', 'Ol

Translation Dictionary Technology Type

In [6]:
tech_type_translation = {
    'Solar': 'Solar',
    'Não Atribuído': 'Not Assigned',
    'Eólica': 'Wind',
    'Biogás': 'Biogas',
    'Cogeração não renovável': 'Non-renewable Cogeneration',
    'Hídrica': 'Hydro',
    'Biomassa': 'Biomass',
    'Fotovoltaica': 'Photovoltaic'
}

df_origin['Technology Type'] = df_origin['Technology Type'].map(tech_type_translation).fillna(df_origin['Technology Type'])
print(df_origin['Technology Type'].unique())


['Solar' 'Not Assigned' 'Wind' 'Biogas' 'Non-renewable Cogeneration'
 'Hydro' 'Biomass' 'Photovoltaic' nan]


### Data Selection

Removing unnecessary columns

In [7]:
cols_to_keep = [
    "Quarter",
    "District",
    "Municipality",
    "Parish",
    "Technology Type",
    "Voltage level",
    "Installed power range (kW)",
    "Number of installations",
    "Total installed power (kW)",
    "CPEs (#)",
    "installations_per_cpe_ratio",
    "power_per_cpe_ratio"
]

df_filtered = df_origin[cols_to_keep]
df_filtered.head(10)

Quarter District     Municipality                              Parish  \
0  2023T1  Coimbra  Condeixa-a-Nova                           Furadouro   
1  2023T1  Coimbra  Condeixa-a-Nova                            Zambujal   
2  2023T1  Coimbra  Condeixa-a-Nova  Condeixa-a-Velha e Condeixa-a-Nova   
3  2023T1  Coimbra  Condeixa-a-Nova               Vila Seca e Bem da Fé   
4  2023T1  Coimbra  Figueira da Foz                           São Pedro   
5  2023T1  Coimbra  Figueira da Foz                  Moinhos da Gândara   
6  2023T1  Coimbra  Figueira da Foz                Buarcos e São Julião   
7  2023T1  Coimbra  Figueira da Foz                Buarcos e São Julião   
8  2023T1  Coimbra  Figueira da Foz                     Ferreira-a-Nova   
9  2023T1  Coimbra  Figueira da Foz                               Lavos   

  Technology Type Voltage level Installed power range (kW)  \
0           Solar           BTN                     ]0, 4]   
1           Solar           BTN                     ]0, 4]   
2    Not Assigned           BTN                     ]0, 4]   
3           Solar           BTN                     ]0, 4]   
4    Not Assigned           BTN                     ]0, 4]   
5    Not Assigned           BTN                     ]0, 4]   
6    Not Assigned           BTN                     ]0, 4]   
7           Solar           BTN                     ]0, 4]   
8    Not Assigned           BTN                     ]0, 4]   
9           Solar           BTN                     ]0, 4]   

   Number of installations  Total installed power (kW)  CPEs (#)  \
0                        2                        3.00    9537.0   
1                        2                        4.32    9537.0   
2                        1                        1.05    9537.0   
3                       17                       28.14    9537.0   
4                        2                        3.28   50436.0   
5                        1                        2.00   50436.0   
6                        9                       13.59   50436.0   
7                      128                      202.51   50436.0   
8                        3                        4.96   50436.0   
9                       93                      112.51   50436.0   

   installations_per_cpe_ratio  power_per_cpe_ratio  
0                     0.000210             0.000315  
1                     0.000210             0.000453  
2                     0.000105             0.000110  
3                     0.001783             0.002951  
4                     0.000040             0.000065  
5                     0.000020             0.000040  
6                     0.000178             0.000269  
7                     0.002538             0.004015  
8                     0.000059             0.000098  
9                     0.001844             0.002231

Selecting years of interest

In [8]:
df_filtered = df_filtered[df_filtered['Quarter'].str.startswith(('2023', '2024'))].copy()
print(df_filtered['Quarter'].value_counts())

Quarter
2023T3    13295
2024T4    11737
2024T3    11399
2024T2    11086
2024T1    10839
2023T1    10481
2023T4    10407
2023T2     9301
Name: count, dtype: int64


### Data Inspection and Cleaning

Checking for missing values

In [9]:
missing_df = pd.DataFrame({
    'Missing Values': df_filtered.isnull().sum(),
    'Percentage': (df_filtered.isnull().sum() / len(df_filtered)) * 100
})
print("\nMissing Values summary:")
display(missing_df[missing_df['Missing Values'] > 0])


Missing Values summary:


Missing Values  Percentage
Voltage level               2    0.002259

In [10]:
print("Before dropping NA:", df_filtered.shape)
df_filtered = df_filtered.dropna()
print("After dropping NA:", df_filtered.shape)

Before dropping NA: (88545, 12)
After dropping NA: (88543, 12)


Check for duplicate rows

In [11]:
duplicates = df_filtered[df_filtered.duplicated()]
print(duplicates)
print("Number of duplicates:", df_filtered.duplicated().sum())

       Quarter  District         Municipality  \
3666    2024T2     Braga             Barcelos   
9221    2024T4    Lisboa                Mafra   
11106   2023T4      Faro                Loulé   
14824   2024T2      Faro                Loulé   
43536   2024T3  Santarém  Salvaterra de Magos   
44840   2023T2     Évora                Évora   
45738   2023T2     Braga            Guimarães   
54983   2024T1      Faro                Loulé   
56600   2024T1    Lisboa               Lisboa   
57315   2023T3     Porto           Matosinhos   
59164   2023T4     Évora                Évora   
62981   2024T2    Lisboa               Loures   
67272   2024T3    Lisboa               Lisboa   
68366   2024T3    Lisboa              Amadora   
87473   2023T4     Braga             Barcelos   
94814   2024T1    Lisboa              Amadora   
96966   2024T2  Santarém  Salvaterra de Magos   
97491   2024T2    Lisboa               Lisboa   
97935   2023T4    Lisboa              Amadora   
99945   2024T1    Li

In [12]:
print("Before dropping duplicates:", df_filtered.shape)
df_filtered = df_filtered.drop_duplicates()
print("After dropping duplicates:", df_filtered.shape)

Before dropping duplicates: (88543, 12)
After dropping duplicates: (88511, 12)


### Derived Metrics

**Percentage of Installations by Voltage Level per District**

According to E-redes ([Manual de Ligações à Rede](https://provedordocliente.e-redes.pt/Files/PDF/Manual-de-Ligacoes-a-Rede.pdf)):

*High Voltage (AT)*
- Companies with capacities >10 MVA, supplied at 60 kV.
- Clear proxy for heavy industry or large commercial facilities.
- Districts with a higher proportion of AT installations → more industrialized.

*Medium Voltage (MT)*
- Companies with capacities >200 kVA, voltages of 10–30 kV.
- Also indicates industrial activity or large commercial companies.
- Complements AT; districts with a higher proportion of MT → more industrialized areas, but less intense than AT.

*Low Voltage Normal (BTN) and Low Voltage Special (BTE)*
- **BTN** → residences, small shops/offices.
- **BTE** → small/medium companies (>41.4 kVA).
- Districts dominated by BTN → mainly residential areas.
- BTE is mixed, can indicate areas with small industries or commerce, but less significant than MT/AT.


In [13]:
# Make a copy
df = df_filtered.copy()

# Group by Quarter, District, and Voltage level, summing the Number of installations
grouped = (
    df.groupby(['Quarter', 'District', 'Voltage level'], as_index=False)['Number of installations'].sum()
)

# Pivot table to have Voltage levels as columns
pivot = grouped.pivot_table(
    index=['Quarter', 'District'],
    columns='Voltage level',
    values='Number of installations',
    fill_value=0
).reset_index()

# Ensure all expected voltage columns exist
for col in ['AT', 'MT', 'BTN', 'BTE']:
    if col not in pivot.columns:
        pivot[col] = 0

# Calculate total installations per row
pivot['Total'] = pivot[['AT','MT','BTN','BTE']].sum(axis=1)

# Calculate percentage per voltage level
pivot["District_High_Voltage_AT(%)"] = pivot['AT'] / pivot['Total'] * 100
pivot["District_Medium_Voltage_MT(%)"] = pivot['MT'] / pivot['Total'] * 100
pivot["District_Low_Voltage_BTN(%)"] = pivot['BTN'] / pivot['Total'] * 100
pivot["District_Low_Voltage_BTE(%)"] = pivot['BTE'] / pivot['Total'] * 100

# Select only the relevant columns
df_result = pivot[['Quarter', 'District',
                   'District_High_Voltage_AT(%)',
                   'District_Medium_Voltage_MT(%)',
                   'District_Low_Voltage_BTN(%)',
                   'District_Low_Voltage_BTE(%)']]

df_result


Voltage level Quarter          District  District_High_Voltage_AT(%)  \
0              2023T1            Aveiro                     0.052687   
1              2023T1              Beja                     0.000000   
2              2023T1             Braga                     0.022525   
3              2023T1          Bragança                     0.000000   
4              2023T1    Castelo Branco                     0.064767   
..                ...               ...                          ...   
139            2024T4           Setúbal                     0.027293   
140            2024T4  Viana do Castelo                     0.014778   
141            2024T4         Vila Real                     0.000000   
142            2024T4             Viseu                     0.016521   
143            2024T4             Évora                     0.000000   

Voltage level  District_Medium_Voltage_MT(%)  District_Low_Voltage_BTN(%)  \
0                                   3.819810                    94.362487   
1                                   3.415560                    92.979127   
2                                   2.201825                    96.564928   
3                                   1.640241                    95.899399   
4                                   2.428756                    95.401554   
..                                       ...                          ...   
139                                 1.305495                    97.643741   
140                                 1.728979                    97.103591   
141                                 1.651020                    97.134995   
142                                 2.602016                    96.241533   
143                                 3.394595                    93.556757   

Voltage level  District_Low_Voltage_BTE(%)  
0                                 1.765016  
1                                 3.605313  
2                                 1.210722  
3                                 2.460361  
4                                 2.104922  
..                                     ...  
139                               1.023472  
140                               1.152653  
141                               1.213985  
142                               1.139931  
143                               3.048649  

[144 rows x 6 columns]

In [14]:
df_final = df_filtered.merge(df_result, on=['Quarter', 'District'], how='left')
df_final.head(10)

Quarter District     Municipality                              Parish  \
0  2023T1  Coimbra  Condeixa-a-Nova                           Furadouro   
1  2023T1  Coimbra  Condeixa-a-Nova                            Zambujal   
2  2023T1  Coimbra  Condeixa-a-Nova  Condeixa-a-Velha e Condeixa-a-Nova   
3  2023T1  Coimbra  Condeixa-a-Nova               Vila Seca e Bem da Fé   
4  2023T1  Coimbra  Figueira da Foz                           São Pedro   
5  2023T1  Coimbra  Figueira da Foz                  Moinhos da Gândara   
6  2023T1  Coimbra  Figueira da Foz                Buarcos e São Julião   
7  2023T1  Coimbra  Figueira da Foz                Buarcos e São Julião   
8  2023T1  Coimbra  Figueira da Foz                     Ferreira-a-Nova   
9  2023T1  Coimbra  Figueira da Foz                               Lavos   

  Technology Type Voltage level Installed power range (kW)  \
0           Solar           BTN                     ]0, 4]   
1           Solar           BTN                     ]0, 4]   
2    Not Assigned           BTN                     ]0, 4]   
3           Solar           BTN                     ]0, 4]   
4    Not Assigned           BTN                     ]0, 4]   
5    Not Assigned           BTN                     ]0, 4]   
6    Not Assigned           BTN                     ]0, 4]   
7           Solar           BTN                     ]0, 4]   
8    Not Assigned           BTN                     ]0, 4]   
9           Solar           BTN                     ]0, 4]   

   Number of installations  Total installed power (kW)  CPEs (#)  \
0                        2                        3.00    9537.0   
1                        2                        4.32    9537.0   
2                        1                        1.05    9537.0   
3                       17                       28.14    9537.0   
4                        2                        3.28   50436.0   
5                        1                        2.00   50436.0   
6                        9                       13.59   50436.0   
7                      128                      202.51   50436.0   
8                        3                        4.96   50436.0   
9                       93                      112.51   50436.0   

   installations_per_cpe_ratio  power_per_cpe_ratio  \
0                     0.000210             0.000315   
1                     0.000210             0.000453   
2                     0.000105             0.000110   
3                     0.001783             0.002951   
4                     0.000040             0.000065   
5                     0.000020             0.000040   
6                     0.000178             0.000269   
7                     0.002538             0.004015   
8                     0.000059             0.000098   
9                     0.001844             0.002231   

   District_High_Voltage_AT(%)  District_Medium_Voltage_MT(%)  \
0                     0.041147                       1.796736   
1                     0.041147                       1.796736   
2                     0.041147                       1.796736   
3                     0.041147                       1.796736   
4                     0.041147                       1.796736   
5                     0.041147                       1.796736   
6                     0.041147                       1.796736   
7                     0.041147                       1.796736   
8                     0.041147                       1.796736   
9                     0.041147                       1.796736   

   District_Low_Voltage_BTN(%)  District_Low_Voltage_BTE(%)  
0                    96.872857                     1.289261  
1                    96.872857                     1.289261  
2                    96.872857                     1.289261  
3                    96.872857                     1.289261  
4                    96.872857                     1.289261  
5                    96.872857                    

**Map Quarters to Seasons**

In [15]:
# Function to convert Quarter to Season
def quarter_to_season(quarter):
    if quarter.endswith('T1') or quarter.endswith('T2'):
        return 'Winter'
    elif quarter.endswith('T3') or quarter.endswith('T4'):
        return 'Summer'
    else:
        return 'Unknown'

# Apply the function to create a new Season column
df_final['Season'] = df_final['Quarter'].apply(quarter_to_season)

# Display first 10 rows
df_final.head(10)


Quarter District     Municipality                              Parish  \
0  2023T1  Coimbra  Condeixa-a-Nova                           Furadouro   
1  2023T1  Coimbra  Condeixa-a-Nova                            Zambujal   
2  2023T1  Coimbra  Condeixa-a-Nova  Condeixa-a-Velha e Condeixa-a-Nova   
3  2023T1  Coimbra  Condeixa-a-Nova               Vila Seca e Bem da Fé   
4  2023T1  Coimbra  Figueira da Foz                           São Pedro   
5  2023T1  Coimbra  Figueira da Foz                  Moinhos da Gândara   
6  2023T1  Coimbra  Figueira da Foz                Buarcos e São Julião   
7  2023T1  Coimbra  Figueira da Foz                Buarcos e São Julião   
8  2023T1  Coimbra  Figueira da Foz                     Ferreira-a-Nova   
9  2023T1  Coimbra  Figueira da Foz                               Lavos   

  Technology Type Voltage level Installed power range (kW)  \
0           Solar           BTN                     ]0, 4]   
1           Solar           BTN                     ]0, 4]   
2    Not Assigned           BTN                     ]0, 4]   
3           Solar           BTN                     ]0, 4]   
4    Not Assigned           BTN                     ]0, 4]   
5    Not Assigned           BTN                     ]0, 4]   
6    Not Assigned           BTN                     ]0, 4]   
7           Solar           BTN                     ]0, 4]   
8    Not Assigned           BTN                     ]0, 4]   
9           Solar           BTN                     ]0, 4]   

   Number of installations  Total installed power (kW)  CPEs (#)  \
0                        2                        3.00    9537.0   
1                        2                        4.32    9537.0   
2                        1                        1.05    9537.0   
3                       17                       28.14    9537.0   
4                        2                        3.28   50436.0   
5                        1                        2.00   50436.0   
6                        9                       13.59   50436.0   
7                      128                      202.51   50436.0   
8                        3                        4.96   50436.0   
9                       93                      112.51   50436.0   

   installations_per_cpe_ratio  power_per_cpe_ratio  \
0                     0.000210             0.000315   
1                     0.000210             0.000453   
2                     0.000105             0.000110   
3                     0.001783             0.002951   
4                     0.000040             0.000065   
5                     0.000020             0.000040   
6                     0.000178             0.000269   
7                     0.002538             0.004015   
8                     0.000059             0.000098   
9                     0.001844             0.002231   

   District_High_Voltage_AT(%)  District_Medium_Voltage_MT(%)  \
0                     0.041147                       1.796736   
1                     0.041147                       1.796736   
2                     0.041147                       1.796736   
3                     0.041147                       1.796736   
4                     0.041147                       1.796736   
5                     0.041147                       1.796736   
6                     0.041147                       1.796736   
7                     0.041147                       1.796736   
8                     0.041147                       1.796736   
9                     0.041147                       1.796736   

   District_Low_Voltage_BTN(%)  District_Low_Voltage_BTE(%)  Season  
0                    96.872857                     1.289261  Winter  
1                    96.872857                     1.289261  Winter  
2                    96.872857                     1.289261  Winter  
3                    96.872857                     1.289261  Winter  
4                    96.872857                     1.289261  Winter  
5 

Check final data types

In [16]:
df_final.dtypes

Quarter                           object
District                          object
Municipality                      object
Parish                            object
Technology Type                   object
Voltage level                     object
Installed power range (kW)        object
Number of installations            int64
Total installed power (kW)       float64
CPEs (#)                         float64
installations_per_cpe_ratio      float64
power_per_cpe_ratio              float64
District_High_Voltage_AT(%)      float64
District_Medium_Voltage_MT(%)    float64
District_Low_Voltage_BTN(%)      float64
District_Low_Voltage_BTE(%)      float64
Season                            object
dtype: object

**Summary of numeric variables**

In [17]:
df = df_final.copy()
df.describe().T

count          mean           std  \
Number of installations        88511.0     17.484414     58.117527   
Total installed power (kW)     88511.0    122.137273    383.756280   
CPEs (#)                       88511.0  45229.553615  60332.903272   
installations_per_cpe_ratio    88511.0      0.000824      0.002353   
power_per_cpe_ratio            88511.0      0.005886      0.031139   
District_High_Voltage_AT(%)    88511.0      0.016606      0.017188   
District_Medium_Voltage_MT(%)  88511.0      2.501376      0.792136   
District_Low_Voltage_BTN(%)    88511.0     95.675196      1.288820   
District_Low_Voltage_BTE(%)    88511.0      1.806821      0.719718   

                                       min           25%           50%  \
Number of installations           1.000000      1.000000      2.000000   
Total installed power (kW)        0.000000     14.050000     30.000000   
CPEs (#)                       1260.000000  10911.000000  27430.000000   
installations_per_cpe_ratio       0.000003      0.000036      0.000117   
power_per_cpe_ratio               0.000000      0.000470      0.001444   
District_High_Voltage_AT(%)       0.000000      0.003992      0.012109   
District_Medium_Voltage_MT(%)     0.893764      1.974736      2.461085   
District_Low_Voltage_BTN(%)      89.988877     95.281230     95.899399   
District_Low_Voltage_BTE(%)       0.805742      1.235248      1.765016   

                                        75%            max  
Number of installations            8.000000    2118.000000  
Total installed power (kW)        82.795000   19600.000000  
CPEs (#)                       57414.000000  399456.000000  
installations_per_cpe_ratio        0.000492       0.052016  
power_per_cpe_ratio                0.004071       3.203661  
District_High_Voltage_AT(%)        0.022525       0.075075  
District_Medium_Voltage_MT(%)      2.685493       4.699666  
District_Low_Voltage_BTN(%)       96.380047      98.293723  
District_Low_Voltage_BTE(%)        2.093719       5.283648

**Number of installations**

Average: 17.47 → on average, each record has ~18 installations.

Standard deviation: 58.11 → high, indicating great variability between municipalities/records.

Distribution: median = 2, 25% = 1, 75% = 8 → asymmetric: most records have few installations, but some have extremely high values (max = 2118), possibly outliers or large production centers.

**Total Installed Power UPAC (kW)**

Average: 122.10 kW → relatively low considering that there are large outliers.

Standard deviation: 383.69 kW → very high, again showing strong variability.

Quartiles: 25% = 14.04 kW, median = 30.00 kW, 75% = 82.79 kW

Maximum: 19,600 kW → indicates the existence of some very large installations.

Conclusion: most units are small, but there are large installations that greatly increase the average.